### ISSUE TITLE
Proposal: Basic expression language for filters inside GraphQL

### ISSUE THREAD

**Issue by ejoebstl** [facebook/graphql#271]
I frequently find myself trying to compose complex filter expressions when working with GraphQL. A frequent use case, for example, is to filter for some number that's lower, or greater than a given parameter.

Supporting expressions, logical as well as arithmetic, would enable a lot of use cases:

```
query getUsers {
  allUsers: allUsers(filter: {age > 23 && name == "Hugo"}) {
     id,
     name
  }
}
```

Pros: 
* Lets you implement proper filtering

Cons: 
* Might conflict with the original purpose of GraphQL
* Definitely complex to define and implement

What I'm currently thinking of doing to get this functionality: 
* Introducing `age_gt` or `name_contains` fields just to support filtering
* Putting expressions in strings and parsing them at server side

I don't like this hacky approaches though and would love to see it implemented in a proper way. 

What's the general opinion on this? If it's positive, I would be happy to start a PR with some formal specifications and use case analysis. 

In my opinion, this could be an optional feature in the GraphQL spec, not that relevant for regular applications, but potentially very important for GraphQL-BAAS services like [graph.cool](https://www.graph.cool/) or [scaphold.io](https://scaphold.io/), which I see as huge enablers for GraphQL. 

Expressions could, of course, also be used on queries, but I'll open another proposal for that feature if the feedback is positive.

**Comment by stubailo** [comment]
What should the filter do? The proposal contains some syntax, but how does one implement the backend of this, to talk to a database, REST API, etc?

**Comment by ejoebstl** [comment]
@stubailo Does the standard have to specify backend implementation of this feature? 
I think the semantics are very clear, but the backend will differ widely, depending which data source (you name it, database, REST, etc.) is used.

**Comment by stubailo** [comment]
> I think the semantics are very clear

I'm not so sure - does this filter on backend fields or GraphQL fields? Like, how do I know that a `User` has an `age` for example? What if that field is present only in the backend database, but isn't a field you can ask for on the GraphQL schema?

I think it's simple for some particular cases where your backend is essentially identical to your schema (which is the case for BaaS).

Personally I think the fact that specific types arguments are not meaningful in the specification is a critical aspect of GraphQL that enables it to be flexible enough to be used with different backends. However, there might be a simpler extension to the language that makes it easier to implement filtering in the backend.

In other words, how is this better than:

```graphql
query getUsers {
  allUsers(filter: "age > 23 && name == 'Hugo'") {
     id,
     name
  }
}
```

Which is already valid GraphQL?

**Comment by ejoebstl** [comment]
> does this filter on backend fields or GraphQL fields?

Excellent question. On GraphQL fields. Exposing database fields would make no sense at all. 

> In other words, how is this better than [...]

I mentioned using expressions in strings as a potential workaround, but that sacrifices type safety!
Integrating a well-defined syntax into the language makes it way easier to check for errors, support syntax highlighting, and test the code.

**Comment by idibidiart** [comment]
Why not specify the filter in the resolver for allUsers with the arguments being $gt and $name? That would keep the business logic outside the schema. 

I'm a purist when it comes to what should go into the GraphQL schema. For me, even operational features like pagination and subscriptions should live in a layer below it.

**Comment by ejoebstl** [comment]
> Why not specify the filter in the resolver for allUsers with the arguments being $gt and $eq?

Good point. I think my proposal makes more sense for BAAS services. 
The key point is that such an API/Extension cannot be leveraged as long as it is not in the standard - since it would break compatibility with existing Frontend libraries.

**Comment by ThisIsMissEm** [comment]
The other thing to consider with the syntax proposed initially is how is it parsed and compiled? What are the rules for grouping operations and evaluating them?

**Comment by ejoebstl** [comment]
> What are the rules for grouping operations and evaluating them?

Great question! I would suggest using left-to-right and braces-first combined with the [operator precedence of Javascript](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Operators/Operator_Precedence), as it has proven to be naturally understandable and is known by a lot of people.

I have to add though, that I would not allow all expressions from the list linked above (infix increment, for example, makes no sense). Also, it might be useful to add some extra functions for date and/or string processing, like (`startsWith`, `contains`, `day`, `hour`) and so on. Such functions calls would have highest precedence, like resolving braces.

**Comment by rmosolgo** [comment]
This sounds like a great application-level convention or even library-level convention, but I think it should be implemented within the current spec using field arguments, as others have suggested above. Using field arguments preserves type safety, which addresses the concern about using a string.

I have a few questions about the suggested change: 

- Filtering based on time: What timezone would be used? 
- How can schema owners enable or disable filtering? 
- Whose responsibility is it to implement efficient filtering? Is it the responsibility of the schema library, or the application developer?

All of these questions can be answered at the application level: 

- Every application has its own approach to handling timezones, it could be applied appropriately 
- In user code, input types could be derived from object types plus a set of configurations -- this would allow schema owners to opt in or out of filtering. 
- Data fetching is an application concern, so by allowing the application to address filtering, you empower the schema owner to implement it as efficiently as they want. 

For this reason, I think the current GraphQL specification is sufficient to support robust filtering. A better solution would be to derive the filter input type from object types, eg 

```js 
var filterConfig = {
  // user-provided filter options 
}

// Generate an input type based on this object type's fields and the user's preferences:
var filterInputType = generateFilterType(objectType, filterConfig) 

// Then, mount it as a `filter:` argument, eg
// users(filter: UsersFilterInput): [User!]!
```

**Comment by sorenbs** [comment]
I really like the proposed syntax. For me this is the main point:

> Integrating a well-defined syntax into the language makes it way easier to check for errors, support syntax highlighting, and test the code (vs a plain string)

At Graphcool we support complex filtering in a type safe way within the current GraphQL specification:

```graphql
query getUsers {
  allUsers: allUsers(filter: {OR:[AND:[{age_gt: 23}, {name: "Hugo"}], {isHappy: true}]}) {
     id
  }
}
```
With this proposal it would be:
```graphql
query getUsers {
  allUsers: allUsers(filter: {(age > 23 && name == "Hugo") || isHappy: true}) {
     id
  }
}
```
With a string this would be:
```graphql
query getUsers {
  allUsers: allUsers(filter: "(age > 23 && name == "Hugo") || isHappy: true") {
     id
  }
}
```

When we designed our filter api we decided that type safety and GraphiQL auto completion was more important than a concise syntax. If we could have both, that would be truly wonderful!

As pointed out in this thread this syntax should defer as many decisions to the schema owner as possible. Operator precedence should probably align with javascript, but obviously this syntax should not prescribe how DateTimes are interpreted or how the application actually resolves a field with an expression argument.

**Comment by langpavel** [comment]
GraphQL should be minimal set of language. Introducing more features will shift it near to SQL. No, for me. GraphQL is next step after REST APIs. It should not replace app domain logic.

**Comment by lunicon** [comment]
Boolean expressions with current object fields will be very useful in directive. For example

```
query getSomeVeryBigObject {
  ObjectType
  SomeData @skip(if: ObjectType != 3) {
    ... data1
  }
  OtherData @skip(if: ObjectType == 3) {
    ... data2
  }  
}
```

**Comment by SanderMertens** [comment]
In addition to type safety, conciseness and syntax coloring, it would provide a uniform method of specifying queries across GraphQL APIs, which IMO would be great for the ecosystem, as it allows for smarter clients & better interoperability.

Perhaps this doesn't need to be added to the language, but there could be a "companion spec" that specifies what the format of a `filter` argument should be, plus some discovery mechanism for finding out which companion specs an API conforms to.

**Comment by gkesler** [comment]
1. What fields/properties are supported in filter expressions? Only immediate properties of the object that contains filter argument or sub-properties are also allowed?
For instance I am interested in your input on the query like below:
```graphql
query getInvoices {
   allInvoices (filter: "amount > 100 && customer.name == "Hugo"") {
     id
     amount
     customer {
       id
       name
     }
   }
}
```
2. Along with regular boolean operations there need to be a set of boolean functions, like `IN`, `BETWEEN`, `LIKE`, etc. How one could specify a function that is not listed in the spec and validation rules?

**Comment by jahudka** [comment]
@langpavel that's why we can't have nice things...

GraphQL is very nice, but if it truly is to be the next REST, it needs to be universal. REST lacks an expressive set manipulation language, opting instead to allow the developer to come up with their own according to their own needs - which is what GraphQL currently does as well. So I can write a REST API that will at the same time allow an external application to integrate fully with my service as well as drive half of the frontend of my service's website. To that end I can come up with my own semantics and conventions for selecting result fields, defining complex filtering and sorting criteria and declaring relations to be loaded alongside the result.. but why do that when there's GraphQL, which does two of those for me out of the box? The issue is that if I want to remain true to GraphQL's nature then just the declaration of complex filter input types is HUGE. Yes, GraphQL may have originated from the specific needs of Facebook ("get posts ordered by ID descending, with like counts and first 10 comments ordered by ID descending" etc.), but there are apps out there that need filtering and sorting by more complex criteria than just IDs..

**Comment by leebyron** [comment]
I'm going to move this proposal to rejected for a few reasons:

* Despite being 18mo old, there has yet to be an actual syntax & semantics proposal
* The primary intent of generic filtering is actively counter to one of GraphQL's guiding principles: that performance is a feature. What would it mean to attempt to "filter" something using this syntax which is not indexed? Or filter a field which is actually a derived value, or provided from a 3rd party service which had rate limiting?
* This syntax adds complexity, but doesn't actively allow for any use case that isn't already possible with input objects. Should we instead be looking at how to better the input type system to better model these use cases?

If this was proposed again, I'd like to see it address all of the reasons above.

**Comment by langpavel** [comment]
@jahudka Sorry, I didn't react on your comment. May be you catch it later.
The point is: GraphQL is universal. Backends behind GQL are various. You don't want maintain filtering on your GraphQL endpoint. You must know how to fetch data from your backing store efficiently.
In most cases — It's not possible to propagate filter logic from GQL to backing store. So GQL cannot support this. Lee is right and correctly closed this.
However there are still possibilities:
- higher level types/resolvers, do you know Relay Connections, right? You can use functional JavaScript to do awesome.
- custom directives with GraphQL extensions, alike Apollo do for tracing
Both solutions are so complex that nobody will even try to write about the idea. Good luck for everyone who will try popularize :sparkles:

**Comment by johncmunson** [comment]
Does anyone know if it's possible to conditionally apply a filter based on the value of a passed in `$argument`?

**Comment by langpavel** [comment]
@johncmunson Just like [Passing Arguments](https://graphql.org/graphql-js/passing-arguments/)

**Comment by johncmunson** [comment]
@langpavel I guess I'm not really following... I wrote up a [SO question](https://stackoverflow.com/questions/53601340/how-can-i-filter-inside-a-graphql-query-based-on-queried-data-and-passed-in-argu) to try and better explain the issue I'm facing, hope that helps

**Comment by langpavel** [comment]
@johncmunson Not related to this discussion, but [I answered you on SO](https://stackoverflow.com/questions/53601340/how-to-conditionally-apply-a-graphql-filter-based-on-the-value-of-a-passed-in-ar/53601831#53601831)

**Comment by ElvenMonky** [comment]
I believe that GraphQL really misses this functionality. GraphQL is great in defining which "columns"/"paths" within each entity server needs to return, but doesn't help at all to define which "rows"/"entities/sub-entities" client wants to get.
Query Language is not complete without this functionality.

I know no query language is perfect, even most comprehensive ones, like MongoDB, SQL, Elastic DSL, etc. are lacking some features from others. But this should not be excuse for GraphQL, but rather an opportunity, because it has a real potential to become Universal Query language, that will be easy to transform into any specific Query DSL.

https://hasura.io/docs/1.0/graphql/core/queries/index.html#queries is a great example how filter, sorting, pagination and aggregation syntax could look like.

Java has QueryDSL, .Net has Entity Framework, but world still needs a cross-language technology to unify access to all data sources. Why not GraphQL?
